cpp/src/arrow/array/array_primitive.cc

```diff
@@ -56,7 +56,7 @@ int64_t BooleanArray::false_count() const {
 }
 
 int64_t BooleanArray::true_count() const {
-  if (data_->null_count.load() != 0) {
+  if (data_->GetNullCount() != 0) {
```

Comment:
Use `data_->MayHaveNulls()` instead. It will do the right thing without triggering the bitmap scan -- that would be extra work that doesn't even speed-up the rest of this function (except for maybe warming up the cache).

https://github.com/apache/arrow/blob/main/cpp/src/arrow/array/data.h#L287-L291

Reply:
Aha, so it's because a boolean check always faster than count?

Reply:
Thanks @felipecrv!

Reply:
> Aha, so it's because a boolean check always faster than count?

@mapleFU `GetNullCount` might scan the entire bitmap if `null_count == kUnknownNullCount`. It's O(length) while `MayHaveNulls()` is a O(1) check.

Reply:
Yeah that I mean it doesn't need to get the exactly count here, just quick check whether it "MayHasNull"